servant/src/Servant/API/Authentication.hs

```diff
@@ -0,0 +1,101 @@
+-- About the type parameters:
+-- m: the monad errors are retrned in. For now just IO.
+-- responseError: an error response type. For now just ServantErr.
+-- policy: the policy to handle OnMissing.
+-- errorIndex: an ADT representing possible errors encountered while extracting
+-- authentication data from a request.
+data OnMissing m responseError (policy :: AuthPolicy) errorIndex where
+    LaxMissing :: OnMissing m responseError 'Lax errorIndex
+    StrictMissing :: (errorIndex -> m responseError) -> OnMissing m responseError 'Strict errorIndex
+
+-- | A GADT indexed by policy strictness that encompasses the ways
+-- users will handle the case where the authentication data provided is
+-- rejected. I.e. a username and password do not match in the database.
+-- About the type parameters:
+-- m: the monad errors are returned in. For now just IO.
+-- responseError: an error response type. For now just ServantErr.
+-- authData: the authentication data extracted from the request
+-- errorIndex: an index of error to give a user-provided, semantic meaning to the authentication failure.
+-- policy: the policy to handle OnUnauthenticated actions.
+data OnUnauthenticated m responseError (policy :: AuthPolicy) errorIndex authData where
+    LaxUnauthenticated :: OnUnauthenticated m e 'Lax errorIndex authData
+    StrictUnauthenticated :: (errorIndex -> authData -> m responseError)
+                          -> OnUnauthenticated m responseError 'Strict errorIndex authData
+
+-- | An ADT representing the data and functions required to protect a reasource for authentication.
+-- For an authenticated resource, we need to handle the scenario where authentication data is missing
+-- and where authentication data is present but not valid (e.g. uesrname + password not valid).
+-- m: the monad errors are retrned in. For now just IO.
+-- rError: an error response type. For now just ServantErr.
+-- mPolicy: Missing Auth: the policy to handle missing authentication data actions.
+-- mError: Missing Auth: ADT error index for possible missing auth failures
+-- uPolicy: Unauthenticated: the policy to handle rejected authentication attempts.
+-- uError: Unauthenticated: ADT error index for possible unauthentication failures
+-- authData: the type of authData present in a request (e.g. JWT token)
+-- usr: a data type extracted from the authenticated data. This data is likely fetched from a database.
+-- subserver: the rest of the servant API.
+data AuthProtected m rError (mPolicy :: AuthPolicy) mError (uPolicy :: AuthPolicy) uError authData usr =
+    AuthProtected { onMissing :: OnMissing m rError mPolicy mError
+                  , onUnauthenticated :: OnUnauthenticated m rError uPolicy uError authData
+                  , checkAuth :: authData -> m (Either uError usr)
+                  }
+
+newtype AuthProtectedSimple req e usr = AuthProtectedSimple { authHandler :: req -> IO (Either e usr) }
```

Comment:
@jkarni here is the `AuthProtectedSimple` that just wraps `Request -> IO (Either ServantErr usr)`. Since we don't have access to `Request` and `ServantErr` in `servant-core` I've just paramaterized them.